File: corpus/cs3/1194-prefixa-infixa-e-posfixa/solutions/a-otima-basica/prefixa.py
import sys

def solve_tree(pre, inf):
    if not pre:
        return ""
        
    root = pre[0]
    pos = inf.find(root)
    
    in_left = inf[:pos]
    in_right = inf[pos+1:]
    
    pre_left = pre[1:pos+1]
    pre_right = pre[pos+1:]
    
    return solve_tree(pre_left, in_left) + solve_tree(pre_right, in_right) + root

def solve():
    input_data = sys.stdin.read().split()
    if not input_data:
        return
        
    C = int(input_data[0])
    idx = 1
    
    out = []
    for _ in range(C):
        N = int(input_data[idx])
        S1 = input_data[idx+1]
        S2 = input_data[idx+2]
        idx += 3
        
        out.append(solve_tree(S1, S2))
        
    print('\n'.join(out))

if __name__ == '__main__':
    solve()

File: corpus/cs3/1194-prefixa-infixa-e-posfixa/solutions/a-otima-basica/prefixa.cpp
#include <iostream>
#include <string>

using namespace std;

string solve(string pre, string in) {
    if (pre.empty()) {
        return "";
    }
    
    char root = pre[0];
    int pos = in.find(root);
    
    string inLeft = in.substr(0, pos);
    string inRight = in.substr(pos + 1);
    
    string preLeft = pre.substr(1, pos);
    string preRight = pre.substr(pos + 1);
    
    return solve(preLeft, inLeft) + solve(preRight, inRight) + root;
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    int C;
    if (!(cin >> C)) return 0;
    
    while (C--) {
        int N;
        string pre, in;
        cin >> N >> pre >> in;
        
        cout << solve(pre, in) << "\n";
    }
    
    return 0;
}
